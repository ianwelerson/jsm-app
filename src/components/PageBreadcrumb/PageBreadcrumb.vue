<script setup lang="ts">
import { computed } from 'vue'
import { useRoute } from 'vue-router'

interface BreadcrumbData {
  routerName: string
  title: string
}

const route = useRoute()

const breadcrumbData = computed(() => {
  const pages: BreadcrumbData[] = [
    {
      routerName: 'appHome',
      title: 'Home',
    },
  ]

  route.matched.reduce((acc, current) => {
    if (!current.meta?.title || !current.name) {
      return acc
    }

    acc.push({
      routerName: current.name as string,
      title: current.meta.title as string,
    })

    return acc
  }, pages)

  return {
    pages,
    total: pages.length,
  }
})
</script>

<template>
  <section v-if="breadcrumbData.total > 1" class="page-breadcrumb">
    <ul class="page-breadcrumb__content">
      <RouterLink
        v-for="(page, index) in breadcrumbData.pages"
        :key="page.routerName"
        class="page-breadcrumb__item"
        :to="{ name: page.routerName }"
        data-testid="breadcrumb-item"
      >
        {{ page.title
        }}<span
          v-if="index + 1 < breadcrumbData.total"
          class="page-breadcrumb__separator"
          >></span
        >
      </RouterLink>
    </ul>
  </section>
</template>

<style lang="scss" scoped>
.page-breadcrumb {
  &__content {
    display: flex;
    align-items: center;
    list-style: none;
    margin: 0;
    padding: 0;
  }

  &__item {
    font-size: $text-xs;
    line-height: $leading-1;
    text-decoration: none;
    color: $text-color;
  }

  &__separator {
    margin: 0 $spacing-1;
  }
}
</style>
