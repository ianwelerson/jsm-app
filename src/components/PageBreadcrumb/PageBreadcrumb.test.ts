import { render } from '@testing-library/vue'
import PageBreadcrumb from '@/components/PageBreadcrumb/PageBreadcrumb.vue'

vi.mock('vue-router', () => ({
  useRoute: vi
    .fn()
    .mockReturnValueOnce({
      matched: [],
    })
    .mockReturnValueOnce({
      matched: [
        {
          name: 'myPage',
          meta: {
            title: 'My Page',
          },
        },
        {
          name: 'mySecondPage',
          meta: {},
        },
      ],
    })
    .mockReturnValue({
      matched: [
        {
          name: 'myPage',
          meta: {
            title: 'My Page',
          },
        },
        {
          name: 'mySecondPage',
          meta: {
            title: 'My Second Page',
          },
        },
      ],
    }),
}))

describe('PageBreadcrumb', () => {
  it('should not render when has only one item', async () => {
    const { queryAllByTestId } = render(PageBreadcrumb)

    expect(queryAllByTestId('breadcrumb-item').length).toBe(0)
  })

  it('should ignore any item without title in meta', async () => {
    const { getByText, queryAllByTestId } = render(PageBreadcrumb)

    getByText('Home')
    getByText('My Page')
    expect(queryAllByTestId('breadcrumb-item').length).toBe(2)
  })

  it('should render all received items', async () => {
    const { getByText, queryAllByTestId } = render(PageBreadcrumb)

    getByText('Home')
    getByText('My Second Page')
    expect(queryAllByTestId('breadcrumb-item').length).toBe(3)
  })

  it('should match the snapshot', async () => {
    const { html } = render(PageBreadcrumb)

    expect(html()).toMatchSnapshot()
  })
})
