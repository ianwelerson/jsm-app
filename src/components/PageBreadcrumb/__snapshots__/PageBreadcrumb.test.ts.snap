// Vitest Snapshot v1

exports[`PageBreadcrumb > should match the snapshot 1`] = `
"<section class=\\"page-breadcrumb\\" data-v-341583bf=\\"\\">
  <ul class=\\"page-breadcrumb__content\\" data-v-341583bf=\\"\\">
    <routerlink class=\\"page-breadcrumb__item\\" to=\\"[object Object]\\" data-testid=\\"breadcrumb-item\\" data-v-341583bf=\\"\\">Home<span class=\\"page-breadcrumb__separator\\" data-v-341583bf=\\"\\">&gt;</span></routerlink>
    <routerlink class=\\"page-breadcrumb__item\\" to=\\"[object Object]\\" data-testid=\\"breadcrumb-item\\" data-v-341583bf=\\"\\">My Page<span class=\\"page-breadcrumb__separator\\" data-v-341583bf=\\"\\">&gt;</span></routerlink>
    <routerlink class=\\"page-breadcrumb__item\\" to=\\"[object Object]\\" data-testid=\\"breadcrumb-item\\" data-v-341583bf=\\"\\">My Second Page
      <!--v-if-->
    </routerlink>
  </ul>
</section>"
`;
